[Bug 19584] Allow target object to be deleted in frontscripts

peter-b opened the pull request.

peter-b committed:
Add std::swap to MCObjectHandle

peter-b committed:
MCS_downloadurl(): Take `MCObjectHandle` as argument

`MCS_downloadurl()` uses `MCObjectHandle` internally, and taking an
`MCObjectHandle` as a target argument allows it to work properly even
if the target object is invalid.

peter-b committed:
[Bug 19584] Allow deletion of target object in frontscripts

If a frontscript deletes the target object, behave exactly as if the
frontscript had consumed the message without passing it.

peter-b committed:
Add MCObjectPartHandle

Add a new movable, copyable subclass of `MCObjectHandle` that
incorporates a part ID.  This is effectively equivalent to
`MCObjectPtr`, with the distinction that it becomes invalidated if the
underlying object is deleted.  It can be converted to an
`MCObjectPtr`.

runrevmark commented on engine/src/externalv1.cpp:
```diff
@@ -1208,12 +1208,10 @@ static MCExternalError MCExternalContextQuery(MCExternalContextQueryTag op, MCEx
-            t_handle = MCtargetptr . object -> GetHandle();
-            if (!t_handle)
+            if (!MCtargetptr)
                 return kMCExternalErrorOutOfMemory;
```
This needs to be kMCExternalErrorObjectDoesNotExist (previously it assumed that targetptr always existed, and thus the only thing that could fail was creation of the weak handle - which always exists now, if the object does).

runrevmark commented:
@livecode-vulcan review ok 7e71ed8

livecode-vulcan commented:
💙 review by @runrevmark ok 7e71ed8905937d657132b14ef4636bd66498042f

livecode-vulcan commented:
😞 test failure 7e71ed8905937d657132b14ef4636bd66498042f

* try-community-armv6-android-api8: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/158)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/160)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/146)
* try-community-universal-ios-iphoneos10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/90)
* try-community-universal-ios-iphonesimulator10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/90)
* try-community-universal-mac-macosx10.6: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/156)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/160)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/160)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/164)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/160)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/164)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/184)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/157)

peter-b committed:
Add constructors to MCObjectPtr

Allow automatic initialisation of `MCObjectPtr` instances.

peter-b committed:
Convert MCtargetptr to a weak reference

Change `MCtargetptr` to be an `MCObjectPartHandle`.  This allows `the
target` to be deleted without leaving `MCtargetptr` dangling.

livecodepanos commented:
@livecode-vulcan review ok f84b490

livecode-vulcan commented:
💙 review by @livecodepanos ok f84b4901af52b0d9675e08cc792b5daeb2dff47e

livecode-vulcan commented:
😞 test failure f84b4901af52b0d9675e08cc792b5daeb2dff47e

* try-community-armv6-android-api8: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/159)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/161)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/147)
* try-community-universal-ios-iphoneos10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/91)
* try-community-universal-ios-iphonesimulator10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/91)
* try-community-universal-mac-macosx10.6: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/157)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/161)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/161)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/165)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/161)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/165)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/185)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/158)

livecodepanos commented on engine/src/mblhandlers.cpp:
```diff
@@ -3526,7 +3526,14 @@ Exec_stat MCHandlePick(void *context, MCParameter *p_parameters)
+		ctxt.LegacyThrow(EE_NOTARGET);
+		t_success = false;
+	}
+	if (t_success)
```
Ah it needs a `{` here, sorry for missing that before :)

livecodepanos commented:
@livecode-vulcan review ok d19f317

livecode-vulcan commented:
💙 review by @livecodepanos ok d19f31736ae4d148e6e66299115fb15be9935652

livecode-vulcan commented:
😞 test failure d19f31736ae4d148e6e66299115fb15be9935652

* try-community-armv6-android-api8: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/162)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/164)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/150)
* try-community-universal-ios-iphoneos10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/94)
* try-community-universal-ios-iphonesimulator10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/94)
* try-community-universal-mac-macosx10.6: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/160)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/164)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/164)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/168)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/164)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/168)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/188)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/161)

livecodepanos commented:
@livecode-vulcan review ok d7aeee2

peter-b commented:
@livecode-vulcan review by @livecodepanos ok 5c8572e

livecode-vulcan commented:
💙 review by @livecodepanos ok 5c8572efd943c6cedf8d206d3441c4806670eb9e

livecode-vulcan commented:
😞 test failure 5c8572efd943c6cedf8d206d3441c4806670eb9e

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/165)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/167)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/153)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/97)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/97)
* try-community-universal-mac-macosx10.6: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/163)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/167)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/167)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/171)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/167)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/171)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/191)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/164)

montegoulding commented on engine/src/object.h:
```diff
@@ -33,6 +33,7 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
 #if 0
 #include <type_traits>
 #endif
+#include <utility>
```
Needs `#include <algorithm>` I believe for pre C++11.
